marshmallow/marshalling.py

```diff
@@ -203,24 +203,26 @@ def deserialize(
             will be ignored.
         :param unknown: Whether to exclude, include, or raise an error for unknown
             fields in the data. Use `EXCLUDE`, `INCLUDE` or `RAISE`.
-        :param type dict_class: Dictionary class used to construct the output.
+        :param type dict_class: Dictionary class used to construct the output when many=False.
+        :param type list_class: List class used to construct the output when many=True.
         :param bool index_errors: Whether to store the index of invalid items in
             ``self.errors`` when ``many=True``.
         :param int index: Index of the item being serialized (for storing errors) if
             serializing a collection, otherwise `None`.
         :return: A dictionary of the deserialized data.
         """
-        if many and data is not None:
+        ret = dict_class() if not many else list_class()
+        if many and data is not None and is_iterable(data):
```

Comment:
Why not use `is_iterable_but_not_string` here?

Reply:
In my understanding, the library regards strings as iterable (which they are) so I didn't want to make an exception here. If you pass string with many=True, marshmallow will iterate over string and pick individual characters as elements, so I think we should act the same way here.

(However, I think it's confusing and weird, because you cannot validate "primitives" with marshmallow anyway (afaik), and accidentally passing in string rather than "actual" iterable such as list should rather raise than pass, but that's just my perspective. I actually have a PR on this in https://github.com/marshmallow-code/marshmallow/pull/622 but there was no clear outcome.)

Reply:
> However, I think it's confusing and weird, because you cannot validate "primitives" with marshmallow anyway (afaik), and accidentally passing in string rather than "actual" iterable such as list should rather raise than pass, but that's just my perspective

Exactly. Passing in a string is invalid usage, so I think raising an error is the correct behavior.

Reply:
> Exactly. Passing in a string is invalid usage, so I think raising an error is the correct behavior.

But marshmallow doesn't currently give error when you pass string as data for `many=True`? Or 2.x at least didn't, was it changed for 3.x?

Reply:
I don't believe 2.x raises an error, but I don't think that should be dispositive. 3.x became more strict in its handling of `None`, so I think it should do the same for strings (and other invalid types).

Reply:
Ok, I'll check that too tomorrow. In a way regarding strings like any iterable makes sense, but it's normally not what the user intended so raising error makes more sense. So we agree on that, nice.

Reply:
@sloria Checked this, marshmallow 3.x still considers a `str` valid for `many=True`:

```python
In [1]: from marshmallow import Schema

In [3]: from marshmallow import fields

In [4]: class Sch(Schema):
   ...:     foo = fields.Int()

In [7]: Sch(many=True).validate([{'foo': 1}, {'foo': 2}])
Out[7]: {}

In [8]: Sch(many=True).validate('asf')
Out[8]: 
{0: {u'_schema': [u'Invalid input type.']},
 1: {u'_schema': [u'Invalid input type.']},
 2: {u'_schema': [u'Invalid input type.']}}
```

Rather than giving out error `{'_schema': ['Invalid input type.']}`.

Could you check #622 (the discussion there, basically). What is your take on it?